task.js

```diff
@@ -0,0 +1,248 @@
+  
+  class Teacher {
+    constructor(students) {
+      this.students =  students;
+  
+      //Створює об'єкт  де ключи це Імя, а значення це масив з обєктів
+      // this.students = students.reduce((group, arr) => {
+      //   const { name } = arr;
+      //   group[name] = group[name] ?? [];
+      //   group[name].push(arr);
+      //   return group;
+      // }, {}); 
+    }
+    getstudents(){
+      console.table( this.students );
+    }
+  
+    // віддає масив імен студентів відсортованих за оцінками
+    getListOfNamesByAverageMark() {
+      let all_average_mark = [];
+      let name_all_average_mark = [];
+  
+      for (let i_name = 0; i_name < this.students.length; i_name++) {   
+        all_average_mark.push({ name : this.students[i_name].name, average_mark : this.students[i_name].getAverageMark()});
+      }
+      // сортуємо
+      all_average_mark.sort(function (x, y) {
+          return x.average_mark - y.average_mark;
+      });
+      // виводимо тільки І'мя
+      all_average_mark.forEach(element => name_all_average_mark.push(element.name));
+      return name_all_average_mark;
+    }
+  
+    // отримати обєкт студента за його ім'ям
+    getStudentByName(name) {  
+      for (let i in this.students) { 
+        if (name == this.students[i].name) {
+            return this.students[i];
+          }   
+      }
+      return "ERROR";
+    }
+  
+    // заміна або видалення знаючи його ім'я
+    removeAndUpdate(name = false, new_object = false) {
+      let index = false
+      for (let i in this.students) { 
+        if (name === this.students[i].name) {
+            index = i ;
+            break;
+          }   
+      } 
+      if (index) {  
+        if (new_object) {
+          this.students.splice(index, 1,new_object);
+        }
+        else {
+          this.students.splice(index, 1);
+        }
+      }
+      else {
+        console.log("There is no such name ------ ");
+      }
+    }
+  
+    // замінити студента на іншого знаючи його ім'я
+    updateStudentByName(new_Student, name) {
+      this.removeAndUpdate(name, new_Student);
+    }
+    // видалити обєкт студента за його ім'ям
+    removeStudentByName(name) {
+      this.removeAndUpdate(name);
+    }
+  
+    //виводить рядок в якому зображено справжні оцінки фейкових учнів
+    findFakeStudent(){
+      let result = ''
+      for (let i_name = 0; i_name < this.students.length; i_name++) {   
+        let averag_marks =  this.students[i_name].marks.reduce((a, b) => (a + b)) / this.students[i_name].marks.length ;
+        if (averag_marks != this.students[i_name].getAverageMark()) 
+          result += `${this.students[i_name].name} - ${this.students[i_name].marks} \n`;
+      }
+  
+      // другий варіант який Орієнтується на назву класа  "FakeStudent"
+      // for (let i_name = 0; i_name < this.students.length; i_name++) {   
+      //   if (this.students[i_name].__proto__.constructor.name == "FakeStudent") 
```

Comment:
It's not protective to use `__proto__`. That's right thought and you can use `instanceof` instead